types/history/index.d.ts

```diff
@@ -27,7 +27,7 @@ export interface History<HistoryLocationState = LocationState> {
 export interface Location<S = LocationState> {
     pathname: Pathname;
     search: Search;
-    state: S;
+    state?: S;
```

Comment:
I think making this optional will possibly break existing users. Having used `history`, the location `state` is kind of an ephemeral type since there's no way to strictly enforce it. It essentially exists to avoid repetitive casts of `location.state` in your own code, and making it optional would necessitate re-introducing those casts (even if it's just by adding postfix-`!`). 

I'm not generally a fan of how `history` specifies the location state since it is unsound (its fairly easy to push a location with state that is not what you expect, regardless of what you put for your type arguments). However, I'm not certain I could justify such a breaking change in a critical package without hearing from one of the definition owners.

In the meantime, you can always avoid using `S` and explicitly cast `state` where needed (the safer approach when appropriately guarded), or add `| undefined` for the location state in your own usages of `Location`/`History`.

Reply:
Thanks for your feedback and explanation!

I ran into this issue when working on my own history implementation trying to comply with the `@types/history` types. Creating a `location` object with a non-optional `state` property when no state is provided isn't possible to do without performing unsafe casts.

I don't think it's a good idea to have type definitions that don't reflect the implementation to avoid repetitive casts or checks for `undefined`. I would expect the type definitions to prevent any runtime errors. A good example is the test that I had to change:
```
let history = createBrowserHistory<{ some: 'state' }>();
history.location.state; // $ExpectType { some: "state"; }
```
I would expect it to be possible to use `history.location.state.some` here but that would result in a runtime error. Now when having a look at the type definitions again I realize that it isn't even possible to provide a `location` or a `state` value to `createBrowserHistory`.

> its fairly easy to push a location with state that is not what you expect, regardless of what you put for your type arguments

I don't know anything about ways to push incorrectly typed states but that sounds much more complex than just accessing the location state before pushing one.

Are there any plans to improve how `history` specifies the location state? I can understand that making a breaking change like this isn't worth the trouble, but it would be nice if there was a plan to make the state handling type safe, either something like this PR or some other change.

Reply:
In my opinion it's better to publish types as strict as possible (at let consumers loosen them by e.g. explicit any casts) than publishing them loose since there's no reliable way to make them stricter in consuming code.

This change is similar to enabling `strictNullChecks` in your project. I'm cautiosly leaning towards favouring this change if `strictNullChecks` is enabled in most projects and if  `state` can actually be `undefined` in `react-router-dom`. Otherwise it might be better to use `History<State | undefined>` in your specific wrapper around `history`.

Reply:
Does that mean you're interested in signing off @eps1lon?

Reply:
>  if `state` can actually be `undefined` in `react-router-dom`. 

That was still unclear to me. Verified in https://codesandbox.io/s/determined-pine-gd95d?file=/src/index.js:224-260 that `history.location.state` can indeed be `undefined`.